src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,244 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+    value_t* matrix_scratch = local_mem.localAcc.get_pointer() + local_range;
+
+    for (index_t c_tile_id = 0; c_tile_id < tile_c_loops; ++c_tile_id) {
+
+      // Extract a matrix block from global memory
+      extract_input_block(matrix_scratch, ndItem, c_tile_id);
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+      index_t mat_index = local_id;
+
+      #pragma unroll
+      for (index_t c_dim_id = 0; c_dim_id < cache_capacity; ++c_dim_id) {
+
+        const index_t vec_index = c_tile_id * cache_capacity + c_dim_id;
+
+        sum += ProductOperator::eval(matrix_scratch[mat_index], local_mem[vec_index]);
+
+        mat_index += local_range + 1; // Adding one as bank offset
+      }
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+    
+    if (nc_dim_index < nc_dim) {
+      lhs_.eval(out_index) = sum;
+    }
+    return sum;
+  } 
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t, bool is_transposed, int cache_line_size, int work_per_thread>
+template <typename ScratchPointerType>
+SYCL_BLAS_INLINE
+void
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::extract_input_block(ScratchPointerType matrix_scratch, cl::sycl::nd_item<1> id, index_t c_tile_id) {
+
+  constexpr int cache_capacity = cache_line_size / sizeof(value_t);
+
+  const index_t local_id = id.get_local_id(0);
+  const index_t group_id = id.get_group(0);
+  const index_t local_range = id.get_local_range(0);
+  const index_t lda = matrix_.getSizeL();
+
+  // Group indices within global grid
+  const index_t nc_group_id = group_id / WGs_per_C_;
+  const index_t c_group_id = group_id % WGs_per_C_;
+
+  // Tile dimensions
+  const int     tile_dim_nc = local_range / cache_capacity;
+  constexpr int tile_dim_c = cache_capacity;
+  
+  // In-tile indices
+  const index_t tile_nc_index = local_id / tile_dim_c;
+  const index_t tile_c_index = local_id % tile_dim_c;
+
+  // Tile-group (local) indices
+  index_t       local_nc_index = tile_nc_index;
+  const index_t local_c_index = tile_c_index + c_tile_id * tile_dim_c;
+  
+  // Global (grid) indices
+  const index_t grid_nc_index = local_nc_index + nc_group_id * local_range;
+  const index_t grid_c_index = local_c_index + c_group_id * local_range;
+
+  // Matrix index calculated based on global grid indices
+  index_t       mat_index = grid_c_index + grid_nc_index * lda;
+
+  // Number of iterations to load all tiles
+  // local_range / tile_dim_nc == cache_capacity
+  constexpr int tile_nc_loops = cache_capacity;
+
+  #pragma unroll
+  for (index_t nc_tile_id = 0; nc_tile_id < tile_nc_loops; ++nc_tile_id) {
+
+    // Bank offset adds 1 word of padding every 'cache_capacity' elements
+    const index_t bank_offset = tile_c_index;
```

Comment:
This could also be moved out of the for loop, or just use the tile_c_index directly.

Reply:
Updated